### PR TITLE
Statically  setup HPM counters

### PR DESCRIPTION
**Pull request by gameboo**

### CODE REVIEW

**Comment by jrtc27** [comment]
> I think a good place for that is in https://github.com/CTSRD-CHERI/riscv-pk/blob/cheri_purecap/machine/encoding.h

That's auto-generated. I think this is fine, they're only ever going to be used in the one function, and it makes it clear what the mapping is.

**Comment by jrtc27** [comment]
@gameboo Have you tested this on anything (preferably QEMU and Flute)? Not for whether the counters actually work in userspace as that requires more effort, but just that existing software continues to boot and you haven't accidentally broken something (which seems unlikely from the code in question if it builds, but always a possibility).

**Comment by heshamelmatary** [comment]
> That's auto-generated. I think this is fine, they're only ever going to be used in the one function, and it makes it clear what the mapping is.

Makes sense. I was also thinking of having them in `riscv-opcodes` so that other projects (e.g., FreeRTOS or RTEMS) can use/pull/update them, but I'm happy just copying that over; I don't think they will change that often.

**Review by heshamelmatary — APPROVED** [review]

**Comment by jrtc27** [comment]
> > That's auto-generated. I think this is fine, they're only ever going to be used in the one function, and it makes it clear what the mapping is.
> 
> Makes sense. I was also thinking of having them in `riscv-opcodes` so that other projects (e.g., FreeRTOS or RTEMS) can use/pull/update them, but I'm happy just copying that over; I don't think they will change that often.

They don't belong there, not until there are standardised ones in the RISC-V spec, IMO.

**Comment by gameboo** [comment]
> @gameboo Have you tested this on anything (preferably QEMU and Flute)? Not for whether the counters actually work in userspace as that requires more effort, but just that existing software continues to boot and you haven't accidentally broken something (which seems unlikely from the code in question if it builds, but always a possibility).

Not yet. Is there something handy I can check?

**Comment by jrtc27** [comment]
> > @gameboo Have you tested this on anything (preferably QEMU and Flute)? Not for whether the counters actually work in userspace as that requires more effort, but just that existing software continues to boot and you haven't accidentally broken something (which seems unlikely from the code in question if it builds, but always a possibility).
> 
> Not yet. Is there something handy I can check?

Easiest would be to grab an MFS_ROOT kernel (CHERI-GFE and CHERI-QEMU-MFS-ROOT, or their -NODEBUG variants, doesn't really matter) from Jenkins (https://ctsrd-build.cl.cam.ac.uk/view/CheriBSD/job/CheriBSD-pipeline/job/dev/424/artifact/artifacts-riscv64-purecap/ is the latest) and try to boot it with the corresponding bbl(-gfe)-riscv64-purecap you've built locally. For QEMU that'd be something like the following (minimal) command line:
```
/local/scratch/aj443/cheri/output/sdk/bin/qemu-system-riscv64cheri -M virt -m 2048 -nographic -bios /path/to/bbl -kernel /path/to/kernel
```
which should boot to a shell you can run cheribsdtest at. For Flute, easiest is to load both the kernel and bbl through GDB (in that order); Jon and Peter are both familiar with that process. It's a little slow compared to netbooting but easier for a one-off thing.

**Comment by gameboo** [comment]
right.

running `# /bin/cheribsdtest-purecap -a` in qemu yields
```
[a lot of stuff with only XFAIL and PASS as AFAICS]
Expected failures:
  test_bounds_varargs_empty_pointer_null: Expected signal 34, got 0
  test_bounds_varargs_vaarg_overflow: Expected signal 34, got 0
  test_bounds_varargs_printf_load: Expected signal 34, got 0
  test_bounds_varargs_printf_store: Expected si_code 1, got 3
  cheribsdtest_vm_shm_open_anon_unix_surprise: tag transfer succeeded
  cheribsdtest_vm_cap_share_fd_kqueue: tag transfer
  cheribsdtest_vm_cap_share_sigaction: tag transfer
SUMMARY: passed 168, expectedly failed 7
```

**Comment by jrtc27** [comment]
Cool, looks good, the stuff you trimmed is only important when there are unexpected failures :)

**Review by jrtc27 — APPROVED** [review]

**Comment by jrtc27** [comment]
Now we just wait for @arichardson to check he's happy

File: machine/minit.c [hunk 1]
```diff
@@ -125,10 +125,196 @@ static void memory_init()
...
+                "csrw mhpmcounter29, t0\n\t"
+                // bitmask in t0
+                "li t0, 0xfffffff8\n\t"
+                // enable user access -- questionable practice here...
+                "csrs mcounteren, t0\n\t"
+                // un-inhibit all counters
+                "csrc mcountinhibit, t0\n\t"
+                // restore exception vector
+                ".align 2\n\t"
+                #if __has_feature(capabilities)
```
Comment (arichardson) [inline machine/minit.c]:
```suggestion
#if __has_feature(capabilities)
```

**Review by arichardson — APPROVED** [review]
Thanks, this LGTM. Not sure if the `#ifdef` indentation is bbl style, if not probably nicer to place them at the start of the line.

**Pull request merged**